libraries/lib-audio-io/AudioIO.cpp

```diff
@@ -2095,63 +2106,86 @@ bool AudioIO::ProcessPlaybackSlices(
             std::fill_n(pointers[i], mProcessingBuffers[bufferIndex].size(), .0f);
          }
 
-         if (pScope) {
-            const auto discardable = pScope->Process(channelGroup, &pointers[0],
-               mScratchPointers.data(),
-               // The single dummy output buffer:
-               mScratchPointers[mNumPlaybackChannels],
-               mNumPlaybackChannels, len);
-            for(int i = 0; i < seq->NChannels(); ++i)
-            {
-               auto& buffer = mProcessingBuffers[bufferIndex + i];
-               buffer.erase(buffer.begin(), buffer.begin() + discardable);
-            }
+         const auto discardable = pScope->Process(channelGroup, &pointers[0],
+            mScratchPointers.data(),
+            // The single dummy output buffer:
+            mScratchPointers[mNumPlaybackChannels],
+            mNumPlaybackChannels, len);
+         for(int i = 0; i < seq->NChannels(); ++i)
+         {
+            auto& buffer = mProcessingBuffers[bufferIndex + i];
+            buffer.erase(buffer.begin(), buffer.begin() + discardable);
          }
+
          bufferIndex += seq->NChannels();
       }
    }
 
+   //samples at the beginning of could have been discarded
+   //in the previous step, proceed with the number of samples
+   //equal to the shortest buffer size available
    auto samplesAvailable = std::min_element(
       mProcessingBuffers.begin(),
       mProcessingBuffers.end(),
       [](auto& first, auto& second) { return first.size() < second.size(); }
    )->size();
 
-   for(auto& buffer : mMasterBuffers)
-      std::fill_n(buffer.begin(), samplesAvailable, 0);
+   //introduced latency may be too high...
+   //but we don't discard what already was written
```

Comment:
I think the answer to your question is in `FillPlayBuffers`. If for some reason `samplesAvailable` will be zeroed, then there will be more processing passes until `GetNeeded() == 0`. Otherwise if I `return false` as you suggest then extra buffer exchange step will be initiated for same frame, which I think is less preferrable.
 May be we should test how it works now with latency values that are much larger than our regular buffer sizes